### ISSUE TITLE
Exception with Range(byte cast from enum)

### ISSUE THREAD

**Issue by MartinZec** [nunit/nunit#3246]
Hello,

We have a test using Range with enum, casted to a byte. 
With NUnit 3.12 I get an exception attached.

Sample code:
```csharp
public enum ENUM1
{
    ZERO= 0x00,
    ONE= 0x01,
    TWO= 0x02
}

[Test, Description("dummy test1")]
public void test1([Range((byte)ENUM1.ZERO, (byte)ENUM1.TWO)] ENUM1 number)
{
// do some dummy testing
}
```

On Testcnetric GUI the test is marked red. If I execute the test with Testcentric I get the attached exception.txt.

Best regards,
Martin

[exception.txt](https://github.com/nunit/nunit/files/3182870/exception.txt)

**Comment by mikkelbu** [comment]
I justed tested this on NUnit-3.8 and there it is possible to generate all the testcases, but each individual testcase fails, like
```
1) Error : RangeEnumAndByte.Class1.test1(0)
System.ArgumentException : Object of type 'System.Int32' cannot be converted to type 'RangeEnumAndByte.Class1+ENUM1'.
   at System.RuntimeType.CheckValue(Object value, Binder binder, CultureInfo culture, BindingFlags invokeAttr)
   at System.Reflection.MethodBase.CheckArguments(Object[] parameters, Binder binder, BindingFlags invokeAttr, CultureInfo culture, Signature sig)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture, Boolean skipVisibilityChecks)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at NUnit.Framework.Internal.Reflect.InvokeMethod(MethodInfo method, Object fixture, Object[] args)
```

On 3.12 it fails in the generation of the testcases
```
1) Invalid : RangeEnumAndByte.Class1
An exception was thrown while loading the test.
System.InvalidOperationException: A value of type System.Int32 (0) cannot be passed to a parameter of type RangeEnumAndByte.Class1+ENUM1.
   at NUnit.Framework.Internal.ParamAttributeTypeConversions.Convert(Object value, Type targetType) in D:\a\1\s\src\NUnitFramework\framework\Internal\ParamAttributeTypeConversions.cs:line 86
   at NUnit.Framework.RangeAttribute.GetData(IParameterInfo parameter) in D:\a\1\s\src\NUnitFramework\framework\Attributes\RangeAttribute.cs:line 175
   at NUnit.Framework.Internal.Builders.ParameterDataSourceProvider.GetDataFor(IParameterInfo parameter) in D:\a\1\s\src\NUnitFramework\framework\Internal\Builders\ParameterDataSourceProvider.cs:line 57
   at NUnit.Framework.CombiningStrategyAttribute.BuildFrom(IMethodInfo method, Test suite) in D:\a\1\s\src\NUnitFramework\framework\Attributes\CombiningStrategyAttribute.cs:line 90
   at NUnit.Framework.Internal.Builders.DefaultTestCaseBuilder.BuildFrom(IMethodInfo method, Test parentSuite) in D:\a\1\s\src\NUnitFramework\framework\Internal\Builders\DefaultTestCaseBuilder.cs:line 116
   at NUnit.Framework.Internal.Builders.NUnitTestFixtureBuilder.BuildTestCase(IMethodInfo method, TestSuite suite) in D:\a\1\s\src\NUnitFramework\framework\Internal\Builders\NUnitTestFixtureBuilder.cs:line 228
   at NUnit.Framework.Internal.Builders.NUnitTestFixtureBuilder.AddTestCasesToFixture(TestFixture fixture, IPreFilter filter) in D:\a\1\s\src\NUnitFramework\framework\Internal\Builders\NUnitTestFixtureBuilder.cs:line 201
   at NUnit.Framework.Internal.Builders.NUnitTestFixtureBuilder.BuildFrom(ITypeInfo typeInfo, IPreFilter filter) in D:\a\1\s\src\NUnitFramework\framework\Internal\Builders\NUnitTestFixtureBuilder.cs:line 81
   at NUnit.Framework.Internal.Builders.DefaultSuiteBuilder.BuildFrom(ITypeInfo typeInfo, IPreFilter filter) in D:\a\1\s\src\NUnitFramework\framework\Internal\Builders\DefaultSuiteBuilder.cs:line 106
```

If you want to generate a test case for each value of an enum I think a better approach would be to use the `TheoryAttribute` instead of the `RangeAttribute`.
```csharp
        [Theory, Description("dummy test1")]
        public void test1(ENUM1 number)
        {
            // do some dummy testing
        }
```

Ps. Why the cast to byte, when your enum is based on an integer?

**Comment by MartinZec** [comment]
I reported the thrown exception because on NUnit 3.10.1 this was working.

I cast to byte because the Range attribute does not accept enums. So NUnit should handle it as a byte and not as an enum. Furthermore, as already mentioned in #3079, for the value "0" the test shows "byte.MinValue" and not "0".

If the first parameter of the range is not starting with 0x00, the test can be generated and executed.
If the first parameter is 0x00, the mentioned exception is thrown.

In other words, if the example would be:

public void test_working([Range((byte)ENUM1.ONE, (byte)ENUM1.TWO)] ENUM1 number)
{
// this test can be executed
}

it would work.

**Comment by CharliePoole** [comment]
Range doesn't except enums because enums are not able to be incremented.

Your method doesn't accept bytes, because it's written to accept enums. You can use ints or bytes and convert to the desired enum inside your test method. But your method must the type that is being passed to it.

**Comment by mikkelbu** [comment]
> I reported the thrown exception because on NUnit 3.10.1 this was working.

I've just tried the examples with 3.10.1.0, and it still fails on my machine.

> I cast to byte because the Range attribute does not accept enums. So NUnit should handle it as a byte and not as an enum. Furthermore, as already mentioned in #3079, for the value "0" the test shows "byte.MinValue" and not "0".

The reason that I asked is that the underlying value of your enum is not a byte, but an integer. Even if we tried to fix the problem then I think that mixing Range and enums are bound to introduce problems. E.g. what if the entries in the enum are not consecutively numbered, like a flag enum
```csharp
[Flags]
public enum Options 
{
    None    = 0,
    Option1 = 1,
    Option2 = 2,
    Option3 = 4,
    Option4 = 8
}
```
Then the range from `None` to `Option4` would have values, like `5`, that does not correspond to an entry in the enum. With respect to the `byte.MinValue` it is still an issue, albeit a very small one IMHO.

> If the first parameter of the range is not starting with 0x00, the test can be generated and executed.
If the first parameter is 0x00, the mentioned exception is thrown.

It still fails on my machine even if we take another number than 0x00 as the start.

Ps. I still think it is easier to use a `Theory` for the enum and you can always filter way unwanted entries using `Assume.That`, see https://github.com/nunit/docs/wiki/Theory-Attribute for more information.

**Comment by mikkelbu** [comment]
And it seems like "translating" `0` to `byte.MinValue` is on purpose.
https://github.com/nunit/nunit/blob/4584f8f5540bba7ee5abb7af3eea7c2b500e002b/src/NUnitFramework/framework/Internal/TestNameGenerator.cs#L386-L392

**Comment by MartinZec** [comment]
Thank you for the input.
As you were mentioning by changing the enum to a byte the test works.

In code:

[Flags]
public enum ENUM1
{
    ZERO= 0x00,
    ONE= 0x01,
    TWO= 0x02
}

not working:

[Test, Description("dummy test1")]
public void test1([Range((byte)ENUM1.ZERO, (byte)ENUM1.TWO)] ENUM1 number)
{
// do some dummy testing
}

working:

[Test, Description("dummy test1")]
public void test1([Range((byte)ENUM1.ZERO, (byte)ENUM1.TWO)] byte number)
{
// do some dummy testing
// using (ENUM1)number
}

Thank you.

**Comment by mikkelbu** [comment]
Glad that it is working for you. I still prefer
```csharp
[Theory, Description("dummy test1")]
public void test2(ENUM1 number)
{
    // Some test
}
```
over
```csharp
[Test, Description("dummy test1")]
public void test1([Range((byte)ENUM1.ZERO, (byte)ENUM1.TWO)] byte number)
{
    // do some dummy testing
    // using (ENUM1)number
}
```

Using the `TheoryAttribute`:
* there are no casts
* the test will still work if more elements are added to the enum without changing the test
* the test will also work for enums where the underlying values of the enum elements are not consecutively numbered
* Uses the enum names in the test name and not the underlying value, i.e. the test cases are named `test2(ONE)`, `test2(TWO)`, and `test2(ZERO)`, and not `test1(1)`, `test1(2)`, and `test1(byte.MinValue)`.

**Comment by CharliePoole** [comment]
@mikkelbu @MartinZec 

An even simpler approach...

```c#
[Test]
public void test2([Values] ENUM1 number)
{
    // Some test
}
```

`ValuesAttribute` supplies all values for an enum if none are given. Does the same for bool as well.
